CLO Refactor

nightdr opened the pull request:
Refactoring ContinuousLocalOptimization to be more flexible for other optimization methods. This version uses the structure I proposed with a separate Optimizer class:
![separated Optimizer class](https://user-images.githubusercontent.com/35091770/170834121-70011284-153c-4890-a674-4793237eed66.png)
rather than keeping the optimizer within the wrapper:
![Optimizer within wrapper](https://user-images.githubusercontent.com/35091770/170834284-46f6302a-faed-42c1-ad0b-06a4df08bb4d.png)
I didn't really have a preference when we discussed this, so if you prefer the latter, let me know and it's pretty easy to swap between the two approaches. The only significant difference between the two I can come up with is that it is easier to swap the optimizer in the former approach. Maybe this is useful if a user wants to have a different objective function/optimization method after a period of evolution?

Changing the interface like this would break existing code, do you think it's worth it to provide a legacy version of the class with some switch between the two if the new version can't be used?

On a side note, the linter is complaining about a lot of older code. So I'll probably have to go through that at some point and change things to get the score above the 9.8 threshold.

nightdr commented:
Travis build seems to be failing because it can't find a proper scipy version [(see here).](https://app.travis-ci.com/github/nasa/bingo/jobs/571711928)

gbomarito commented:
Also, I'm ok with breaking backwards compatibility at this point (since we are pre v1.0)

And lets ignore travis (we should probably remove it from the CI because we're using github actions now)

nightdr commented on bingo/local_optimizers/optimizer.py:
```diff
@@ -0,0 +1,44 @@
+        individual : `Chromosome`
+            The individual whose constants will be optimized.
+        """
+        raise NotImplementedError
```
That sounds reasonable to me, I like the idea of being able to use something like a lambda function for simple optimization.

Do you think the local optimization fitness function should be responsible for setting up self._optimization_function() (similar to the second approach in the diagrams, options would have to be passed along for setup) or should the object be provided with an already configured optimization_function() (similar to the first approach in the diagrams)?

nightdr replied:
Sounds good with me, what we have should cover that. I'll note the lambda case in either the class' docstring or the doc pages

nightdr commented on bingo/local_optimizers/scipy_optimizer.py:
```diff
@@ -0,0 +1,281 @@
+            self._sub_routine_for_obj_fn, individual, c_0)
+        individual.set_local_optimization_params(params)
+
+    def _sub_routine_for_obj_fn(self, params, individual):
```
I agree that the private functions shouldn't be published to the public docs. I added documentation here for the sake of easier development; although it's a bit overkill. I could replace the full docstrings with short descriptions for functions that aren't described in full by their name.

gbomarito replied:
yes, i think i prefer this style for private functions because we might be more inclined to update them in the future if they are simple/short

nightdr replied:
sounds good with me, I will update those then

nightdr commented on bingo/local_optimizers/continuous_local_opt.py:
```diff
@@ -162,104 +56,24 @@ def eval_count(self):
+            self.optimizer(individual)
         return self._fitness_function(individual)
 
 
```
I think we could combine the interface into ```Chromosome```. If that were the case, I would want to add a ```params``` property to ```Chromosome```. However, to me, the interface suggests there exists a case where we don't want to optimize every parameter. In this case, it would be nice to have the interface and I would see merit to not combining it with ```Chromosome``` since it's possible not every ```Chromosome``` with parameters would care about optimization.

An alternative that maybe you were referencing is we could include the interface in ```Chromosome``` but set it up to the normal case where every parameter would be optimized. This way, the case where not every parameter is optimized can still be configured via ```get_number_local_optimization_params``` and ```set_local_optimization_params```, but not every ```Chromosome``` has to implement them. If we want to go with this approach, should we keep ```needs_local_optimization``` abstract? Maybe it can turn into a property (e.g. ```needs_opt``` or similar)?

nightdr replied:
Decided on renaming ```ChromosomeInterface``` to ```LocalOptimizationInterface``` and moving it to its own file. We also decided that ```Chromosome``` should implement ```LocalOptimizationInterface``` but raise a ```NotImplementedError``` in the interface methods by default.

I couldn't think of a significant difference between ```needs_local_optimization``` and a ```needs_opt``` property, so I left the ```needs_local_optimization``` method in the interface. Let me know if you can think of any benefits to the property approach.

nightdr commented on bingo/local_optimizers/optimizer.py:
```diff
@@ -0,0 +1,44 @@
+from abc import ABCMeta, abstractmethod
+
+
+class OptimizerBase(metaclass=ABCMeta):
```
sounds good to me

tylertownsend commented on bingo/local_optimizers/local_opt.py:
```diff
@@ -0,0 +1,115 @@
+        return self._fitness_function(individual)
+
+
+class ChromosomeInterface(metaclass=ABCMeta):
```
Maybe we should consider moving the `ChromosomeInterface` to its own file since its an interface.

nightdr replied:
I agree and have done so!

nightdr commented:
Renamed the optimization interface, moved it to its own file, and made ```Chromosome``` implement it (see the conversation about the interface for more details). Also renamed some of the modules to align with their classes.

I decided to leave the changes in doc/ out of this PR since I'll be looking at the docs in the near future.

gbomarito commented on bingo/chromosomes/chromosome.py:
```diff
@@ -93,6 +94,36 @@ def distance(self, chromosome):
+        bool
+            Whether `Chromosome` needs optimization
+        """
+        raise NotImplementedError
```
```suggestion
        raise NotImplementedError("This Chromosome cannot be used in local optimization until its local optimization interface has been implemented")
```

gbomarito commented on bingo/chromosomes/chromosome.py:
```diff
@@ -8,30 +8,31 @@
 import copy
 from abc import ABCMeta, abstractmethod
 
+from ..local_optimizers.local_opt_interface import LocalOptimizationInterface
```
I think we can probably get rid of this separate interface file, since the base class already has dummy versions of the functions.

what do you think?

gbomarito commented on bingo/local_optimizers/local_opt_interface.py:
```diff
@@ -0,0 +1,44 @@
+"""This module contains the abstract definition of an interface
+for `Chromosome`s that want to use local optimization.
```
See above comment on whether this file is needed

gbomarito reviewed (changes requested):
looks solid, only one real comment/question

nightdr commented on tests/unit/local_optimizers/test_scipy_optimizer.py:
```diff
@@ -0,0 +1,309 @@
+
+    dummy_individual = DummyLocalOptIndividual()
+
+    opt_options = {"method": method,  # TODO have to set this or else it will get overwritten by BFGS on update, I don't like this
```
Decided on keeping this behavior but documenting it.
